torch/optim/optimizer.py

```diff
@@ -175,6 +176,25 @@ def zero_grad(self):
                         p.grad.requires_grad_(False)
                     p.grad.zero_()
 
+    def reset_grad(self):
+        r"""Set the gradient of of all optimized :class:`torch.Tensor` s to None.
+        This is DIFFERENT from zero_grad, which sets all gradients to 0s instead
+        of None. This will in general have lower memory footprint, but using this
+        comes with caveats, to name a few:
+        1. When user tries to access the gradient value and perform manual ops on it.
+          A None attribute or a Tensor full of 0s will be different.
+        2. User can no longer rely on checking if `.grad` is None to see if a tensor
+          is touched in the backward pass
+        3. `nn.optim` optimizers have a different behavior if the gradient is 0 or None
+          (in one case it does the step with a gradient of 0 and in the other it skip
+          the step altogether).
+        """
+        for group in self.param_groups:
+            for p in group['params']:
+                if p.grad is not None:
+                    p.grad.detach_()
```

Comment:
Why do we need to `detach_()`?  Afaict the only way this has an effect is if something external also holds a reference to `.grad` _and_ `.grad` was created in a `create_graph=True` backward pass.

Reply:
My point is `detach_` may be the _unsafe_ option.  Because it affects `p.grad` in place, it silently affects anything else holding a reference to `p.grad`.  Setting `p.grad = None` without `detach_` simply drops our reference.  Anything else holding a reference to `grad` will not see an effect.

Admittedly, I'm not sure why the default zeroing behavior of `zero_grad` performs `detach_` then `zero_`.  I assume it's to avoid building up spurious autograd history if the grad was created with `create_graph=True` and therefore requires grad.

If so, in the alternative set-to-None path, we dont need `detach_`.  We drop the reference to `.grad`, we don't perform any ops on it, therefore there's no danger of building spurious autograd history.  And if we don't `detach_`, we don't risk silently affecting other references to `grad`.

tl;dr I think _not_ detaching is the safe option here.

@albanD you're good with these tricky cases...Why is `detach_` used with the default zeroing behavior of `zero_grad`?  Also, do you agree _not_ detaching is the safer implementation for the set-to-None path?

Reply:
The original code is trying to make sure that we don't change the Tensor referenced by `.grad` as much as possible I think.
And yes as mentioned below, I do agree that we don't want to detach here.